src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -2566,32 +2566,44 @@ public int ChildCount()
         }
 
         // Determines whether the node supports a compilation / code generation strategy based on walking the node tree.
-        internal bool SupportsCompilation()
+        // Also returns a human-readable string to explain the reason to the source generator.
+        internal bool SupportsCompilation([NotNullWhen(false)] out string? reason)
         {
             if (!StackHelper.TryEnsureSufficientExecutionStack())
             {
-                // If we can't recur further, code generation isn't supported as the tree is too deep.
+                reason = "the regex is too deep";
                 return false;
             }
 
-            if ((Options & (RegexOptions.RightToLeft | RegexOptions.NonBacktracking)) != 0)
+            // NonBacktracking isn't supported, nor RightToLeft.  The latter applies to both the top-level
+            // options as well as when used to specify positive and negative lookbehinds.
+            switch (Options & (RegexOptions.RightToLeft | RegexOptions.NonBacktracking))
             {
-                // NonBacktracking isn't supported, nor RightToLeft.  The latter applies to both the top-level
-                // options as well as when used to specify positive and negative lookbehinds.
-                return false;
+                case 0:
+                    break;
+                case RegexOptions.RightToLeft:
+                    reason = "RegexOptions.RightToLeft is specified";
+                    return false;
+                case RegexOptions.NonBacktracking:
+                    reason = "RegexOptions.NonBacktracking is specified";
+                    return false;
+                default:
+                    reason = "RegexOptions.RightToLeft and RegexOptions.NonBacktracking are specified";
+                    return false;
```

Comment:
A switch isn't going to be appropriate here.  We should first check for NonBacktracking and use that for the error if it exists, and then if it doesn't check for RightToLeft.